src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers/AddConfigureAwaitAnalyzer.cs

```diff
@@ -0,0 +1,131 @@
+        }
+
+        private void CompilationStart(CompilationStartAnalysisContext context) 
+        {
+            _asyncUtilities = new AsyncAnalyzerUtilities(context.Compilation);
+            context.RegisterSyntaxNodeAction(AnalyzeAwaitExpression, SyntaxKind.AwaitExpression);
+            context.RegisterSyntaxNodeAction(AnalyzeUsingExpression, SyntaxKind.UsingStatement);
+            context.RegisterSyntaxNodeAction(AnalyzeForEachExpression, SyntaxKind.ForEachStatement);
+            context.RegisterSyntaxNodeAction(AnalyzeConfigureAwaitTrue, SyntaxKind.InvocationExpression);
+        }
+
+        private void AnalyzeAwaitExpression(SyntaxNodeAnalysisContext context)
+        {
+            var invocation = (AwaitExpressionSyntax)context.Node;
+            var operation = context.SemanticModel.GetOperation(invocation, context.CancellationToken);
+            if (!(operation is IAwaitOperation awaitOperation)) {
+                return;
+            }
+
+            if (_asyncUtilities.IsTaskType(awaitOperation.Operation.Type)) 
+            {
+                ReportConfigureAwaitDiagnostic(context, awaitOperation);
+            }
+        }
+
+        private void AnalyzeUsingExpression(SyntaxNodeAnalysisContext context) 
+        {
+            if (!(context.SemanticModel.GetOperation(context.Node, context.CancellationToken) is IUsingOperation usingOperation))
+            {
+                return;
+            }
+
+            if (!usingOperation.IsAsynchronous)
+            {
+                return;
+            }
+
+            var resources = usingOperation.Resources;
+            if (_asyncUtilities.IsAsyncDisposableType(resources.Type))
+            {
+                ReportConfigureAwaitDiagnostic(context, resources);
+            }
+        }
+
+        private void AnalyzeForEachExpression(SyntaxNodeAnalysisContext context) 
+        {
+            if (!(context.SemanticModel.GetOperation(context.Node, context.CancellationToken) is IForEachLoopOperation forEachOperation))
+            {
+                return;
+            }
+
+            if (!forEachOperation.IsAsynchronous)
+            {
+                return;
+            }
+
+            var collectionType = forEachOperation.Collection.Type;
+            if (_asyncUtilities.IsAsyncEnumerableType(collectionType)) 
+            {
+                ReportConfigureAwaitDiagnostic(context, forEachOperation.Collection);
+            }
+        }
+
+        private void AnalyzeConfigureAwaitTrue(SyntaxNodeAnalysisContext context)
+        {
+            if (!_asyncUtilities.IsConfigureAwait(context.Node))
+            {
+                return;
+            }
+
+            if (!(context.SemanticModel.GetOperation(context.Node, context.CancellationToken) is IInvocationOperation operation)) 
+            {
+                return;
+            }
+
+            if (!_asyncUtilities.IsConfigureAwait(operation.TargetMethod)) 
```

Comment:
What is the difference between this check and the one above where we just check the .Node?

Reply:
This is a complete check but requires semantic model which takes some time to obtain. Previous check validates on the level of a syntax tree, which is quicker but isn't sufficient.

Reply:
Makes sense!